lib/node-loader-core.js

```diff
@@ -61,26 +65,47 @@ function die(msg, err) {
   process.exit(1);
 }
 
-export const resolve = createHook("resolve");
-
-export const getFormat = createHook("getFormat");
-
-export const getSource = createHook("getSource");
-
-export const transformSource = createHook("transformSource");
-
-// getGlobalPreloadCode is synchronous, which means we can't import the config file in time :'(
-// export const getGlobalPreloadCode = createHook("getGlobalPreloadCode")
+/**
+ * Return a supported ESM loader hook appropriate for export.
+ *
+ * @param {"resolve"|"getFormat"|"getSource"|"transformSource"} name - Valid ESM loader hook name
+ * @returns {LoaderHook} - The final composed definition for hook type "name"
+ */
+function createHook(name) {
+  return async function (...args) {
+    if (loadingConfig) {
+      const defaultImpl = args[args.length - 1];
+      return defaultImpl(...args);
+    } else {
+      const config = await getConfigPromise();
+      const { resolvedLoaders } = config;
+      const hook = resolvedLoaders[name];
+      return hook(...args);
+    }
+  };
+}
 
-global.nodeLoader = global.nodeLoader || {};
-global.nodeLoader.setConfigPromise = function setConfigPromise(newConfig) {
+/**
+ * Directly supply a config (no config file) to produce a
+ * resolved loaders definition. Used only for tests.
+ *
+ * @param {ConfigPromise} newConfig
+ */
+function setConfigPromise(newConfig) {
   configPromise = newConfig.then((config) => {
     config.resolvedLoaders = resolveLoaders(config);
     return config;
   });
-};
+}
 
+/**
+ * Used to await async import and resolution of the config.
+ *
+ * @returns {Promise<LoaderDefinition>}
+ */
 function getConfigPromise() {
+  const configPath = getConfigPath();
```

Comment:
This change to the implementation allows for the possibility of the node loader config being changed dynamically at runtime by changing the `process.env.NODE_LOADER_CONFIG` env variaable. I didn't want to allow that behavior when I originally implemented node-loader-core, as I thought it would be confusing to have your loaders start behaving differently in the middle of program execution.

Did you refactor the implementation to specifically allow for this? Or was it just an artifact of the refactor?

Reply:
This was primarily to make [testing easier](https://github.com/node-loader/node-loader-core/pull/4/files#diff-57f954ce0fc8810478e5b870750102e3f8efee746769a1d8e0c811438a4c4467R97). Without a way to programmatically get the config path, the variable would need to be set in the npm `test` script definition. I'm happy to switch to that approach, as I can see why you wouldn't want this possibility.

Reply:
I think it's fine - seems unlikely people will change the `NODE_LOADER_CONFIG` env variable during execution.